pkg/utils/subst/subst.go

```diff
@@ -36,110 +42,156 @@ func ParseFile(file string, fss ...vfs.FileSystem) (SubstitutionTarget, error) {
 	return s, nil
 }
 
+var yqLibLogInit sync.Once
+
 func Parse(data []byte) (SubstitutionTarget, error) {
+	yqLibLogInit.Do(func() {
+		var lvl glog.Level
+		switch ocmlog.Context().GetDefaultLevel() {
+		case mlog.None:
+			fallthrough
+		case mlog.ErrorLevel:
+			lvl = glog.ERROR
+		case mlog.WarnLevel:
+			lvl = glog.WARNING
+		case mlog.InfoLevel:
+			lvl = glog.INFO
+		case mlog.DebugLevel:
+			fallthrough
+		case mlog.TraceLevel:
+			lvl = glog.DEBUG
+		}
+		glog.SetLevel(lvl, "yq-lib")
+	})
+
 	var (
-		err     error
-		content interface{}
-		fi      fileinfo
+		err error
+		fi  fileinfo
 	)
 
 	fi.json = true
-	if err = runtime.DefaultJSONEncoding.Unmarshal(data, &content); err != nil {
+	rdr := bytes.NewBuffer(data)
+	jsnDcdr := yqlib.NewJSONDecoder()
+
+	if err := jsnDcdr.Init(rdr); err != nil {
+		return nil, err
+	}
+
+	if fi.content, err = jsnDcdr.Decode(); err != nil {
 		fi.json = false
-		if err = runtime.DefaultYAMLEncoding.Unmarshal(data, &content); err != nil {
-			return nil, errors.Wrapf(err, "no yaml or json data")
+		ymlPrfs := yqlib.NewDefaultYamlPreferences()
+		ymlDcdr := yqlib.NewYamlDecoder(ymlPrfs)
+		rdr = bytes.NewBuffer(data)
+		if err := ymlDcdr.Init(rdr); err != nil {
+			return nil, err
+		}
+		if fi.content, err = ymlDcdr.Decode(); err != nil {
+			return nil, err
 		}
-		data, err = runtime.DefaultYAMLEncoding.Marshal(content)
-	} else {
-		data, err = runtime.DefaultJSONEncoding.Marshal(content)
-	}
-	if err != nil {
-		return nil, errors.Wrapf(err, "cannor marshal data")
 	}
-	// mixed json/yaml cannot be parsed, modified and marshalled again, correctly
-	// so try to come with pure yaml or pure json.
 
-	fi.content, err = parser.ParseBytes(data, 0)
-	if err != nil {
-		return nil, errors.Wrapf(err, "invalid YAML")
-	}
+	fi.content.SetDocument(0)
+	fi.content.SetFilename("substitution-target")
+	fi.content.SetFileIndex(0)
```

Comment:
Why not simply use `NewAllAtOnceEvaluator` and evulate the nodes instead of doing SetDocument and the likes? I think we can completely remove the need for using / passing around the AST. We can just take in something and actually produce the output as a `[]byte`.

Reply:
I wouldn't want to remove the caching of the AST in this PR for fear of having a negative impact on performance.  I say this because 
- I see ocm-controller caching SubstitutionTargets during mutation ( so someone has some perf concerns already )
- Similarly SubstitutionTargets were caching the AST 
- I know that in CSI we will be making several updates to the same SubstitutionTarget, so if parsing and reparsing is slow it will impact us
- We're already doing 1 thing that could have a performance impact, changing the lib, and if we make another change, taking out the cache, and someone complains we won't know which change was the problem
- I haven't seen any performance tests that would tell us that removing the AST caching is safe

So again, in my opinion if taking out the caching is desirable for some reason I think that should be a separate PR and that PR should also add some performance tests.

Oh, and a non-performance concern.... not having the cached object makes debugging more difficult if you are having multiple updates happening to the same substitution target because in `Substitute(subs Substitutions, fs vfs.FileSystem) ` your won't be able to 
- You won't be able to add a watch that lets you quickly see what each of the rules are doing to the SubstitutionTarget
- You won't be able to add a conditional breakpoint that lets you catch which rule is making a particular change to a SubstitutionTarget

Reply:
Fair point 👍

Reply:
Right, but those are for subs rules. That would happen once with the other parser as well. 🤔

Anyways, I played around with it a bit and actually don't really like it, because it fucks up the multiline formatting. It adds that pesky little `|-` to everything. :D And it's so annoying. So I don't care. :D Leave it be like that. :)

Reply:
Also, the JSON part is completely broken somehow and it further ignores a couple of proper indentations. It's really weird, that it works nicely, once you get into AST level. :D So I think what you are doing is better.